File: packages/android-sdk/gateflow/src/main/java/com/gateflow/sdk/logger/Logger.kt
package com.gateflow.sdk.logger

import android.util.Log

/** Centralized logger with severity levels and scoped categories. */
object Logger {
    private const val TAG = "GateFlow"

    var level: LogLevel = LogLevel.INFO
    var delegate: LogDelegate? = null

    @JvmStatic
    fun debug(
        level: LogLevel = LogLevel.DEBUG,
        scope: LogScope = LogScope.GateFlowCore,
        message: String,
        info: Map<String, Any>? = null,
        error: Throwable? = null,
    ) {
        if (level.priority < this.level.priority) return

        val infoStr = info?.entries?.joinToString(prefix = " | ", separator = ", ") { "${it.key}=${it.value}" } ?: ""
        val logMsg = "[${scope.name}] $message$infoStr"

        when (level) {
            LogLevel.DEBUG -> Log.d(TAG, logMsg, error)
            LogLevel.INFO -> Log.i(TAG, logMsg, error)
            LogLevel.WARN -> Log.w(TAG, logMsg, error)
            LogLevel.ERROR -> Log.e(TAG, logMsg, error)
        }

        delegate?.handleLog(level, scope, message, info, error)
    }
}

/** Receives SDK log events for forwarding to app-level logging. */
interface LogDelegate {
    fun handleLog(level: LogLevel, scope: LogScope, message: String, info: Map<String, Any>?, error: Throwable?)
}

File: packages/ios-sdk/Sources/GateFlowKit/Logger/Logger.swift
import Foundation

/// Centralized logger with severity levels and scoped categories.
enum GFLogger {
    static var level: LogLevel = .info
    static weak var delegate: LogDelegate?

    private static let subsystem = "GateFlow"

    static func debug(
        level: LogLevel,
        scope: LogScope,
        message: String,
        info: [String: Any]? = nil,
        error: Error? = nil
    ) {
        guard level.priority >= self.level.priority else { return }

        let infoStr = info.map { " | \($0)" } ?? ""
        let errorStr = error.map { " | \($0)" } ?? ""
        let full = "[\(subsystem)][\(scope.rawValue)] \(message)\(infoStr)\(errorStr)"

        switch level {
        case .debug: print(full)
        case .info: print(full)
        case .warn: print(full)
        case .error: print(full)
        }

        delegate?.handleLog(level: level, scope: scope, message: message, info: info, error: error)
    }
}

/// Receives SDK log events for forwarding to app-level logging.
protocol LogDelegate: AnyObject {
    func handleLog(level: LogLevel, scope: LogScope, message: String, info: [String: Any]?, error: Error?)
}
